src/wasm/kotlin/java/arcs/Entity.kt

```diff
@@ -13,131 +13,163 @@ package arcs
-        val token = str.substring(0, len)
-        str = str.substring(len)
-        return token
+        val chunk = bytes.sliceArray(0..(len - 1))
+        bytes = bytes.sliceArray(len..(bytes.size - 1))
+        return chunk
     }
 
     fun validate(token: String) {
-        if (chomp(token.length) != token) {
+        if (chomp(token.length).decodeToString() != token) {
             throw Exception("Packaged entity decoding failed in validate()\n")
         }
     }
 
     fun decodeText(): String {
-        val len = getInt(":")
-        return chomp(len)
+        val len = getInt(':')
+        return chomp(len).decodeToString()
     }
 
     fun decodeNum(): Double {
-        val token = upTo(":")
-        return token.toDouble()
+        return upTo(':').decodeToString().toDouble()
     }
 
     fun decodeBool(): Boolean {
-        return (chomp(1)[0] == '1')
+        return (chomp(1).decodeToString() == "1")
     }
 }
 
-class StringEncoder(private val sb: StringBuilder = StringBuilder()) {
-
-    companion object {
-        fun encodeDictionary(dict: Map<String, Any?>): String {
-            val sb = StringBuilder()
-            sb.append(dict.size).append(":")
+class StringEncoder(private val buffers: MutableList<ByteArray> = ArrayList<ByteArray>(),
+                    private var size: Int = 0) {
 
-            for ((key, value) in dict) {
-                sb.append(key.length).append(":").append(key)
-                sb.append(encodeValue(value))
-            }
-            return sb.toString()
+    fun encodeDictionary(dict: Map<String, Any?>): StringEncoder {
+        addStr("${dict.size}:")
+        for ((key, value) in dict) {
+            addBytes("", key.encodeToByteArray())
+            encodeValue(value)
         }
+        return this
+    }
 
-        fun encodeList(list: List<Any>): String {
-            return list.joinToString(separator = "", prefix = "${list.size}:") { encodeValue(it) }
-        }
+    fun encodeList(list: List<Any>): StringEncoder {
+        addStr("${list.size}:")
+        list.forEach { encodeValue(it) }
+        return this
+    }
 
-        fun encodeValue(value: Any?): String {
-            return when (value) {
-                is String -> "T${value.length}:$value"
-                is Boolean -> "B${if (value) 1 else 0}"
-                is Double -> "N$value:"
-                is Map<*, *> -> {
-                    @Suppress("UNCHECKED_CAST")
-                    val dictString = encodeDictionary(value as Map<String, Any?>)
-                    "D${dictString.length}:$dictString"
-                }
-                is List<*> -> {
-                    @Suppress("UNCHECKED_CAST")
-                    val listString = encodeList(value as List<Any>)
-                    "A${listString.length}:$listString"
-                }
-                else -> throw IllegalArgumentException("Unknown expression.")
+    private fun encodeValue(value: Any?) {
+        return when (value) {
```

Comment:
This could also probably be an expression function:

```kotlin
private fun encodeValue(value: Any?) = when (value) {
    //...
}
```

Reply:
fyi this was due to a different problem: `encodeValue` shouldn't be returning any more. Small oversight when I changed this to use underlying byte construction.